propulate/propulator.py

```diff
@@ -261,10 +274,32 @@ def _evaluate_individual(self) -> None:
         """Breed and evaluate individual."""
         ind = self._breed()  # Breed new individual.
         start_time = time.time()  # Start evaluation timer.
-        if self.worker_sub_comm != MPI.COMM_SELF:
-            ind.loss = self.loss_fn(ind, self.worker_sub_comm)  # Evaluate its loss.
+
+        # signal start of run to surrogate model
+        if self.surrogate is not None:
+            self.surrogate.start_run(ind)
+
+        # define local loss_fn for parallelized evaluation
+        def loss_fn(individual):
+            if self.worker_sub_comm != MPI.COMM_SELF:
+                return self.loss_fn(individual, self.worker_sub_comm)
+            else:
+                return self.loss_fn(individual)
```

Comment:
Just asking out of curiosity (bc I am not a computer scientist): Why is it better to have this as a function definition here instead of a mere if-else block (as before):
```
        if self.worker_sub_comm != MPI.COMM_SELF:
            ind.loss = self.loss_fn(ind, self.worker_sub_comm)  # Evaluate its loss.
        else:
            ind.loss = self.loss_fn(ind)
```
I find your way more elegant but was told that function calls in Python would be slow.

Reply:
It's to avoid deeper nesting when adding the `isgeneratorfunction()` check and it helps with readability as the two different if -  for parallelized workers and generator loss function - are separated.

The alternative without any extra function calls would be something along the lines
```python
        if self.worker_sub_comm != MPI.COMM_SELF:
            if inspect.isgeneratorfunction(self.loss_fn):
                last: float = 0.0
                for last in self.loss_fn(ind, self.worker_sub_comm):
                    if self.surrogate is not None:
                        if self.surrogate.cancel(last):  # Check cancel for each yield.
                            break
                ind.loss = last  # Set final loss as individual's loss.
            else:
                ind.loss = self.loss_fn(ind, self.worker_sub_comm)  # Evaluate its loss.
        else:
            if inspect.isgeneratorfunction(self.loss_fn):
                last: float = 0.0
                for last in self.loss_fn(ind):
                    if self.surrogate is not None:
                        if self.surrogate.cancel(last):  # Check cancel for each yield.
                            break
                ind.loss = last  # Set final loss as individual's loss.
            else:
                ind.loss = self.loss_fn(ind)  # Evaluate its loss.
```
Which is not only less readable but also introduces a lot of code repetition. 

Python function calls are _relatively_ slow. As long as `_evaluate_individual()` isn't called thousands of times per second per process, the overhead should be negligible and is well worth it for better readability imo.

Reply:
There's also a shorter version with lambdas:
```python
        if self.worker_sub_comm != MPI.COMM_SELF:
            loss_fn = lambda ind: self.loss_fn(ind, self.worker_sub_comm)
        else:
            loss_fn = lambda ind: self.loss_fn(ind)
```
But that's more a matter of taste rather than a functional improvement.

Reply:
Thanks for the clarification, we will leave it as is with the function definition.